src/transformers/models/groupvit/modeling_tf_groupvit.py

```diff
@@ -0,0 +1,2013 @@
+logger = logging.get_logger(__name__)
+
+# soft dependency
+if is_tensorflow_probability_available():
+    try:
+        import tensorflow_probability as tfp
+
+        # On the first call, check whether a compatible version of TensorFlow is installed
+        # TensorFlow Probability depends on a recent stable release of TensorFlow
+        n = tfp.distributions.Normal(loc=0.0, scale=1.0)
+    except ImportError:
+        logger.error(
+            "GroupViT models are not usable since `tensorflow_probability` can't be loaded."
+            "It seems you have `tensorflow_probability` installed with the wrong tensorflow version."
+            "Please try to reinstall it following the instructions here: https://github.com/tensorflow/probability."
+        )
+
+_CHECKPOINT_FOR_DOC = "nvidia/groupvit-gcc-yfcc"
+
+TF_GROUPVIT_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "nvidia/groupvit-gcc-yfcc",
+    # See all GroupViT models at https://huggingface.co/models?filter=groupvit
+]
+
+
+################################################################################
+# Utility Functions
+################################################################################
+
+
+LARGE_NEGATIVE = -1e8
+
+
+# Copied from transformers.models.bart.modeling_tf_bart._expand_mask
+def _expand_mask(mask: tf.Tensor, tgt_len: Optional[int] = None):
+    """
+    Expands attention_mask from `[bsz, seq_len]` to `[bsz, 1, tgt_seq_len, src_seq_len]`.
+    """
+    src_len = shape_list(mask)[1]
+    tgt_len = tgt_len if tgt_len is not None else src_len
+    one_cst = tf.constant(1.0)
+    mask = tf.cast(mask, dtype=one_cst.dtype)
+    expanded_mask = tf.tile(mask[:, None, None, :], (1, 1, tgt_len, 1))
+
+    return (one_cst - expanded_mask) * LARGE_NEGATIVE
+
+
+# contrastive loss function, adapted from
+# https://sachinruk.github.io/blog/pytorch/pytorch%20lightning/loss%20function/gpu/2021/03/07/CLIP.html
+def contrastive_loss(logits: tf.Tensor) -> tf.Tensor:
+    return tf.math.reduce_mean(
+        tf.keras.metrics.sparse_categorical_crossentropy(
+            y_true=tf.range(shape_list(logits)[0]), y_pred=logits, from_logits=True
+        )
+    )
+
+
+# Copied from transformers.models.clip.modeling_tf_clip.clip_loss with clip->groupvit
+def groupvit_loss(similarity: tf.Tensor) -> tf.Tensor:
+    caption_loss = contrastive_loss(similarity)
+    image_loss = contrastive_loss(tf.transpose(similarity))
+    return (caption_loss + image_loss) / 2.0
+
+
+def hard_softmax(logits: tf.Tensor, dim: int) -> tf.Tensor:
+    """
+    Reference: https://gist.github.com/ariG23498/08cdae21637b8b61bdd6d21d11719fb3
+    """
+    y_soft = stable_softmax(logits, dim)
+    # Straight through.
+    index = tf.argmax(y_soft, dim)
+    y_hard = tf.one_hot(
+        index,
+        depth=shape_list(logits)[dim],
+        # TensorFlow expects axis to be -1 or between [0, 3).  But received: -2
+        # This is why the following code snippet is used.
+        axis=range(len(shape_list(logits)))[dim],
```

Comment:
Hey @gante this fails if the dim is a -ve integer. Here `dim=-2`.

Reply:
Using `tf.rank` and making it more generic, I am using `axis=range(tf.rank(logits))[dim]`. WDYT?

Reply:
ah, you're right -- we can use `tf.math.maximum(dim, tf.rank(logits) - dim)`, which will work for the positive and negative cases

(I'd like to avoid solutions based on `range` or `tf.range`, as they create arrays and will necessarily be slower)

Reply:
I don't think this will work either.

Let's consider the rank of logits to be 4. With `dim=-2`, the equation simplifies to
```
tf.math.maximum(-2, 4-(-2)) = tf.math.maximum(-2, 6) = 6
```
This again fails as the correct axis.

Reply:
Hehe you're right, this actually needs a mod operator!

With `tf.math.floormod(dim, tf.rank(logits))` we get the correct number :)

Reply:
This is such a nice op. Loved this!

Reply:
I had tried using this but the tests were failing as in mentioned in this [comment](https://github.com/huggingface/transformers/pull/18020#issuecomment-1231146285).